### PR TITLE
Enable defining custom test environments

### PR DESCRIPTION
**Pull request by andrew4699**
# Description

This is a test-only change which adds a pattern for running tests against user-specified environments by specifying your own base URL and HTTP client. Makes PolarisRestCatalogViewIntegrationTest use the custom test environment. In a future PR I plan to do similar for other tests but I'd like this PR to just start the pattern.

Example:
```
INTEGRATION_TEST_BASE_URL="http://localhost:8181" \
INTEGRATION_TEST_HTTP_CLIENT_FACTORY_IMPL="org.apache.polaris.MyCustomHttpClientFactory" \
INTEGRATION_TEST_ROLE_ARN="arn:aws:iam::<role>" \
INTEGRATION_TEST_S3_PATH="s3://<bucket>" \
./gradlew test --info --rerun --tests "org.apache.polaris.service.catalog.PolarisRestCatalogViewIntegrationTest"
```

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [X] Existing tests work
- [X] Tested running against a non-Dropwizard Polaris with a custom HTTP client

# Checklist:

- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings

### CODE REVIEW

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironment.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.test;
+
+import jakarta.ws.rs.client.Client;
+import java.net.URI;
+
+/** Defines the test environment that a test should run in. */
+public class TestEnvironment {
```
Comment (eric-maynard) [inline polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironment.java:25]: Can this be a `record?

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+   * to the root path of your Dropwizard instance.
+   */
+  public static final String ENV_BASE_URL = "INTEGRATION_TEST_BASE_URL";
+
+  /**
+   * Environment variable that specifies the HTTP client factory that should construct test HTTP
+   * clients. This should be a class name. If this is not set, falls back to the client created by
+   * Dropwizard.
+   */
+  public static final String ENV_HTTP_CLIENT_FACTORY_IMPL =
```
Comment (eric-maynard) [inline polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java:47]: This approach seems a bit roundabout. Is there a way we can achieve this with a service file?

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    TestHttpClientFactory httpClientFactory;
+    try {
+      httpClientFactory =
+          (TestHttpClientFactory)
+              (Class.forName(httpClientImpl.get()).getDeclaredConstructor().newInstance());
+    } catch (Exception e) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Cannot initialize http Client, %s does not implement PolarisTestHttpClientFactory.",
```
Comment (eric-maynard) [inline polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java:109]: It looks like this is now named `TestHttpClientFactory`. Also, let's not catch `Exception`

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  private static Client getHttpClient(DropwizardAppExtension dropwizardAppExtension) {
+    Optional<String> httpClientImpl =
+        Optional.ofNullable(System.getenv(ENV_HTTP_CLIENT_FACTORY_IMPL));
+
+    if (httpClientImpl.isEmpty()) {
+      if (dropwizardAppExtension == null) {
+        throw new ParameterResolutionException(
+            "Tried to default to Dropwizard client but could not find DropwizardAppExtension.");
+      }
+      return dropwizardAppExtension.client();
```
Comment (eric-maynard) [inline polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java:98]: Should this logic actually get encapsulated in (e.g.) `DropWizardTestHttpClientFactory`?

**Review by eric-maynard — CHANGES_REQUESTED** [review]
Left a few questions

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironment.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * under the License.
+ */
+package org.apache.polaris.service.test;
+
+import jakarta.ws.rs.client.Client;
+import java.net.URI;
+
+/** Defines the test environment that a test should run in. */
+public record TestEnvironment(Client apiClient, String baseUrl) {
+  public URI getBaseUri() {
```
Comment (collado-mike) [inline polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironment.java:26]: You could make the main constructor take a URI and have a secondary constructor that takes the string and converts to a URI. That way the conversion is done on construction and always done once.

**Review by collado-mike — APPROVED** [review]

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java (hunk 4 above)
Comment (andrew4699) [reply polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java]: Updated to TestEnvironmentResolver + DropwizardTestEnvironmentResolver

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java (hunk 3 above)
Comment (andrew4699) [reply polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java]: Updated

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironment.java (hunk 1 above)
Comment (andrew4699) [reply polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironment.java]: Updated

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironment.java (hunk 5 above)
Comment (andrew4699) [reply polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironment.java]: Updated

File: polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java (hunk 2 above)
Comment (andrew4699) [reply polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java]: For things like this the project seems to have relied on providing a test config file and Jackson parsing the config. I'm hoping to decouple this from Dropwizard and for non-Dropwizard environments also not force people to go through that whole dance.

**Review by eric-maynard — APPROVED** [review]
LGTM, thanks!

**Review by eric-maynard — APPROVED** [review]

**Pull request merged**